### PR TITLE
C++: Adding a model implementation for ODBC.

### PR DESCRIPTION
**Pull request by bdrodes**
Adds a model implementation for ODBC database APIs.

### CODE REVIEW

File: cpp/ql/lib/semmle/code/cpp/models/Models.qll [hunk 1]
```diff
@@ -9,6 +9,7 @@ private import implementations.Iterator
 private import implementations.MemberFunction
 private import implementations.Memcpy
 private import implementations.Memset
+private import implementations.ODBC
```
Comment (MathiasVP) [inline cpp/ql/lib/semmle/code/cpp/models/Models.qll:12]: Maybe we should move it down to the other SQL-like imports?

File: cpp/ql/lib/semmle/code/cpp/models/implementations/ODBC.qll [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+ *
+ * Note, `SQLExecute` is not included because it operates on a SQLHSTMT type, not a string.
+ * The SQLHSTMT parameter for `SQLExecute` is set through a `SQLPrepare`, which is modeled.
+ * The other source of input to a `SQLExecute` is via a `SQLBindParameter`, which sanitizes user input,
+ * and would be considered a barrier to SQL injection.
+ */
+private class ODBCExecutionFunction extends SqlExecutionFunction {
+  ODBCExecutionFunction() {
+    exists(string s | s in ["SQLExecDirect", "SQLPrepare"] and this.hasName(s))
+  }
```
Comment (MathiasVP) [inline cpp/ql/lib/semmle/code/cpp/models/implementations/ODBC.qll:22]: Two changes:
- No need to introduce a `string`. Let's just inline the set literal (which is what we call the `[a, b, c]` syntax).
- Replace `hasName` with `hasGlobalName`. I don't think we want to match member functions named `SQLExecDirect` or `SQLPrepare`, right? The two functions we're interested are non-member functions in the global namespace, right?

```suggestion
  ODBCExecutionFunction() { this.hasGlobalName(["SQLExecDirect", "SQLPrepare"]) }
```

**Review by MathiasVP — COMMENTED** [review]
A couple of comments.

Additionally, would it be possible to add some tests to the end of the https://github.com/github/codeql/blob/main/cpp/ql/test/query-tests/Security/CWE/CWE-089/SqlTainted/test.c file? Or if you prefer to the C++ file: https://github.com/github/codeql/blob/main/cpp/ql/test/query-tests/Security/CWE/CWE-089/SqlTainted/test.cpp

Basically, we just want a test that confirms that we can track tainted flow coming into the sink.

**Comment by MathiasVP** [comment]
DCA looks good: performance is unaffected, and we're finding 4 new results that seem genuine. So I think this PR is good to go once the above comments have been addressed 🎉

**Comment by bdrodes** [comment]
Thanks. I'll look over the change suggestions shortly.

File: cpp/ql/lib/semmle/code/cpp/models/Models.qll (hunk 1 above)
Comment (bdrodes) [reply cpp/ql/lib/semmle/code/cpp/models/Models.qll]: Done

File: cpp/ql/lib/semmle/code/cpp/models/implementations/ODBC.qll (hunk 2 above)
Comment (bdrodes) [reply cpp/ql/lib/semmle/code/cpp/models/implementations/ODBC.qll]: Done

**Comment by bdrodes** [comment]
@MathiasVP, I've made the suggested changes and added a quick ODBC sql injection test. Let me know if there is anything else need to close this out. Thanks

**Comment by MathiasVP** [comment]
Looks good. Thanks!

The only thing we need now is to make the autoformatter happy. CI tells me that the `ODBC.qll` file needs to be [formatted correctly](https://github.com/github/codeql/blob/main/docs/ql-style-guide.md).

You can do this using `codeql query format -i path-to-the-file` to format it in-place. Alternatively, you can do it via the `Format Document` command in VSCode.

**Comment by bdrodes** [comment]
> Looks good. Thanks!
> 
> The only thing we need now is to make the autoformatter happy. CI tells me that the `ODBC.qll` file needs to be [formatted correctly](https://github.com/github/codeql/blob/main/docs/ql-style-guide.md).
> 
> You can do this using `codeql query format -i path-to-the-file` to format it in-place. Alternatively, you can do it via the `Format Document` command in VSCode.

Whoops. I did this on my initial push but forgot to do it for the recent changes. It should be all good now.

**Review by MathiasVP — APPROVED** [review]
LGTM!

**Pull request merged**